src/io/procfs.rs

```diff
@@ -306,3 +328,151 @@ type StaticFd = OnceCell<(OwnedFd, Stat)>;
+/// mounted on top of it, and that it looks normal.
+///
+/// # References
+///  - [Linux]
+///
+/// [Linux]: https://man7.org/linux/man-pages/man5/proc.5.html
+#[inline]
+pub fn proc_self_fdinfo_fd<Fd: AsFd>(fd: &Fd) -> io::Result<OwnedFd> {
+    let fd = fd.as_fd();
+    _proc_self_fdinfo(fd)
+}
+
+fn _proc_self_fdinfo(fd: BorrowedFd<'_>) -> io::Result<OwnedFd> {
+    let (proc_self_fdinfo, proc_self_fdinfo_stat) = proc_self_fdinfo()?;
+    let fd_str = DecInt::from_fd(&fd);
+    open_and_check_file(proc_self_fdinfo, proc_self_fdinfo_stat, fd_str.as_z_str())
+}
+
+/// Returns a handle to a Linux `/proc/self/pagemap` file.
+///
+/// This ensures that `/proc/self/pagemap` is `procfs`, that nothing is
+/// mounted on top of it, and that it looks normal.
+///
+/// # References
+///  - [Linux]
+///  - [Linux pagemap]
+///
+/// [Linux]: https://man7.org/linux/man-pages/man5/proc.5.html
+/// [Linux pagemap]: https://www.kernel.org/doc/Documentation/vm/pagemap.txt
+#[inline]
+pub fn proc_self_pagemap() -> io::Result<OwnedFd> {
+    proc_self_file(zstr!("pagemap"))
+}
+
+/// Returns a handle to a Linux `/proc/self/maps` file.
+///
+/// This ensures that `/proc/self/maps` is `procfs`, that nothing is
+/// mounted on top of it, and that it looks normal.
+///
+/// # References
+///  - [Linux]
+///
+/// [Linux]: https://man7.org/linux/man-pages/man5/proc.5.html
+#[inline]
+pub fn proc_self_maps() -> io::Result<OwnedFd> {
+    proc_self_file(zstr!("maps"))
+}
+
+/// Open a file under `/proc/self`.
+fn proc_self_file(name: &ZStr) -> io::Result<OwnedFd> {
+    let (proc_self, proc_self_stat) = proc_self()?;
+    open_and_check_file(proc_self, proc_self_stat, name)
+}
+
+/// Open a procfs file within in `dir` and check it for bind mounts.
+fn open_and_check_file(dir: BorrowedFd, dir_stat: &Stat, name: &ZStr) -> io::Result<OwnedFd> {
+    let (_, proc_stat) = proc()?;
+
+    let oflags =
+        OFlags::RDONLY | OFlags::CLOEXEC | OFlags::NOFOLLOW | OFlags::NOCTTY | OFlags::NOATIME;
+    let file = openat(&dir, name, oflags, Mode::empty()).map_err(|_err| io::Error::NOTSUP)?;
+    let file_stat = fstat(&file)?;
+
+    // Open a copy of the `dir` handle so that we can read from it
+    // without modifying the current position of the original handle.
+    let dot = openat(&dir, zstr!("."), oflags | OFlags::DIRECTORY, Mode::empty())
+        .map_err(|_err| io::Error::NOTSUP)?;
+
+    // Confirm that we got the same inode.
+    let dot_stat = fstat(&dot).map_err(|_err| io::Error::NOTSUP)?;
+    if (dot_stat.st_dev, dot_stat.st_ino) != (dir_stat.st_dev, dir_stat.st_ino) {
+        return Err(io::Error::NOTSUP);
+    }
+
+    // `is_mountpoint` only works on directory mount points, not file mount
+    // points. To detect file mount points, scan the parent directory to see
+    // if we can find a regular file with an inode and name that matches the
+    // file we just opened. If we can't find it, there could be a file bind
+    // mount on top of the file we want.
```

Comment:
Is there a real world attack that resulted from something creating a hostile bind mount over `/proc`?  I struggle to think of a case where malicious code could create a bind mount over e.g. `/proc/<pid>/fdinfo` or something but not also have the capability to create the same mount over e.g. `/lib/ld-linux.so.2` and have compromised the process from the start too.

(The answer here is probably better put in the doc comments at the top I guess)

Reply:
I don't know of any real-world attacks. It's mainly about how people are doing all kinds of different things with Linux containers these days, with all kinds of different blends of what we once called "root" and "user" privileges, so in general we want to avoid as many vectors for misdirection as possible. I've added some docs.

For a hypothetical example, processes which launch outside the sandbox and then use `chroot` or other things to enter it wouldn't be vulnerable to `/lib/ld-linux.so.2` inside the sandbox being redirected. Such programs may need to be careful about trusting filesystem paths once they've entered the sandbox, but rustix and especially cap-std will sometimes use procfs implicitly, in ways that aren't documented or obvious from the public API, which could make it tricky to avoid any such trust. With the checking here, rustix and its users may fail in some cases, but they shouldn't be able to succeed with bogus data.

Reply:
OK.  I think though if you have `openat()` there is no reason to use `chroot()`.  `chroot()` is really a dangerous thing to use if you potentially have *anything* in your address space which isn't explicitly expecting it.

(Tangential but interesting side fact; One of the rationales for https://github.com/coreos/rpm-ostree/ avoiding using the [rpm shared library](https://github.com/rpm-software-management/rpm/) is actually because it's a systemd service offering a DBus API, and librpm since the very first days has used `chroot()` to run scripts.  rpm-ostree reimplements all that, creating a container for each script.  We also use `openat()` and will be using cap-std in more places)

IOW I think cap-std should really be thought of as a replacement for `chroot()`.

I'm not going to argue against the extra checks here, I just wanted to understand.  The new comment seems fine.

Reply:
Maybe it's useful: systemd [uses statx](https://github.com/systemd/systemd/blob/main/src/basic/mountpoint-util.c#L193) for this.

Reply:
Thanks for asking questions! This has been helpful for me. I tend to agree about `chroot`, so I myself probably wouldn't write code that way, but I'm aiming to have rustix be ok with someone else doing it.

Also thanks for that systemd link; I didn't know about `STATX_ATTR_MOUNT_ROOT`. That looks new in Linux 5.8 so while we may use that to optimize, we'd still need the fallback code. I've added a comment about it.